src/StaticLS/HIE/File.hs

```diff
@@ -0,0 +1,122 @@
+module StaticLS.HIE.File where
+
+import Control.Monad
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.IO.Unlift
+import Control.Monad.Trans (lift)
+import Control.Monad.Trans.Maybe (MaybeT (..), runMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask)
+import Data.List.Extra (dropPrefix)
+import Data.Maybe (catMaybes, mapMaybe)
+import qualified Data.Text as T
+import Data.List 
+import HieDb
+import StaticLS.Monad
+import System.Directory (makeAbsolute, doesFileExist)
+import System.FilePath (normalise, (</>), (-<.>))
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified GHC
+import qualified GHC.Iface.Ext.Binary as GHC
+import qualified GHC.Iface.Ext.Types as GHC
+import qualified GHC.Types.Name.Cache as GHC
+import qualified GHC.Unit.Types as GHC
+import qualified Language.LSP.Types as LSP
+
+type SrcFilePath = FilePath
+type HieFilePath = FilePath
+data HieInfo = HieInfo
+    { hieFilePath :: HieFilePath
+    , hieFile :: GHC.HieFile
+    }
+
+-- TODO: make this configurable (use hie.yaml?)
+hieDir :: FilePath
+hieDir = ".hiefiles/"
+
+-- TODO: make this configurable (use cabal?)
+srcDirs :: [FilePath]
+srcDirs = ["src/", "lib/", "test/"]
+
+-- | Retrieve a hie info from a lsp text document identifier
+getHieFile :: HasStaticEnv m => LSP.TextDocumentIdentifier -> m (Maybe GHC.HieFile)
+getHieFile tdi = do
+    staticEnv <- getStaticEnv
+    hieFilePath <- runMaybeT $ do
+      srcFilePath <- MaybeT $ pure $ LSP.uriToFilePath tdi._uri
+      MaybeT $ srcFilePathToHieFilePath srcFilePath
+    liftIO $ mapM (pure . GHC.hie_file_result <=< GHC.readHieFile staticEnv.nameCache) hieFilePath
+
+-----------------------------------------------------------------------------------
+-- File/Directory method for getting hie files - faster but somewhat "hacky"
+-----------------------------------------------------------------------------------
+
+hieFilePathToSrcFilePath :: HasStaticEnv m => HieFilePath -> m SrcFilePath
+hieFilePathToSrcFilePath hiePath = do
+  staticEnv <- getStaticEnv
+  -- TODO: handle failure here
+  hieFileResult <- liftIO $ GHC.readHieFile staticEnv.nameCache hiePath
+  liftIO $ makeAbsolute hieFileResult.hie_file_result.hie_hs_file
+
+-- | Retrieve a hie file path from a src path
+--
+-- Substitutes the src directory with the hie directory and the src file extension with
+-- the hie file extension. Fragile, but works well in practice.
+--
+-- Presently necessary because hiedb does not currently index the hs_src file location
+-- in the `mods` table
+srcFilePathToHieFilePath :: HasStaticEnv m => SrcFilePath -> m (Maybe HieFilePath)
```

Comment:
Inspired by halfsp's method but relatively more robust (halfsp's method does a naiive find a replace of `.hs` and the src dirs while this checks prefixes/only replaces the file extension)

Reply:
i think a more reliable way to do this is to use `cabal info .` to get the list of modules and then turn those module names into the corresponding paths.  then you don't need to hardcode the source directories.